docs/blog/2023-07-15-incremental.md

```diff
@@ -0,0 +1,226 @@
+---
+slug: Changes
+title: Computing with changes
+authors: [mbudiu]
+tags: [incremental computation, changes, deltas, transaction, diff, database, CDC]
+unlisted: true
+---
+
+# Incremental computation
+
+*Incremental computation* refers to computations that attempt to
+reuse results that were previously computed.
+
+Let us assume you have a program *P* that is given an input *I* and
+produces an output *O*:
+
+![computation](computation.svg)
+
+Now let's change the input by *modifying* it with some change, which
+we denote by *ΔI*:
+
+![reuse](reuse.svg)
+
+We would like to avoid computing the entire output again, and ideally
+reuse the previously computed *O* and only adjust it by a change *ΔO*.
+This would be great especially if the output doesn't *change* much.
+
+This sounds nice, but is very vague.  To make this work we have to
+explain what a __change__ is (what we denoted by Δ) and what it means
+to change something (what we denoted by +).  For general-purpose
+computations this in general makes no sense, but it works very well in
+some important domains.
+
+## Changes in documents
+
+If you have used [change
+tracking](https://support.microsoft.com/en-us/office/track-changes-in-word-197ba630-0f5f-4a8e-9a77-3712475e806a)
+in Microsoft Word, you have encountered a version of changes.  Here is
+a screenshot of Word showing changes to short document:
+
+![word document](word.PNG)
+
+Word keeps track of the changes by looking at your keystrokes and
+recording what is being deleted (shown with a ~~strikethrough~~) or
+inserted (shown as underlined).  Notice that Word doesn't attempt to
+show the minimal changes performed -- which could be shown just by the
+deletion of the apostrophe `'` in *it's*, but it rather mirrors the
+actions you have performed to change the document.
+
+Change tracking is such a useful feature in Word because it allows
+readers to avoid reading a whole edited document to discover what has
+changed.  In general
+
+> change tracking is mostly useful when changes are expected to be
+  *small* compared to the modified object
+
+## Changes in programs
+
+Most programmers have encountered a natural notion of change: since
+programs evolve constantly, one often speaks about the changes between
+two versions of a program.  Important activities of programmers, such
+as [code reviews](https://en.wikipedia.org/wiki/Code_review), often
+only look at changes in programs.
+
+There are specific tools, like
+[diff](https://en.wikipedia.org/wiki/Diff) which compute changes
+between text files.  Here is an example from the above Wikipedia page:
+
+```console
+$ diff original new
+0a1,6
+> This is an important
+> notice! It should
+> therefore be located at
+> the beginning of this
+> document!
+>
+11,15d16
+< This paragraph contains
+< text that is outdated.
+< It will be deleted in the
+< near future.
+17c18
+< check this dokument. On
+---
+> check this document. On
+```
```

Comment:
No one seriously uses the original and obsolete `diff` output format. It's unreadable and unsafe. Unified (`-u`) diffs are the standard these days. I don't think we should try to promote this obsolete format.

Reply:
It's the first time I hear about this. I only use the obsolete format.
But we are not promoting anything, we just describe how changes surface in other contexts.

Reply:
I'm astonished.

Reply:
So, can I give the diff example?
I don't use diff with tools, just for human consumption.

Reply:
It's OK, I guess, I'm just surprised that anyone still tolerates such an awful format.

Reply:
The quote here comes directly from the wikipedia page, I didn't make it myself.